pkg/volume/downwardapi/downwardapi.go

```diff
@@ -0,0 +1,329 @@
+	volName                 string
+	fieldReferenceFileNames map[string]string
+	pod                     *api.Pod
+	plugin                  *downwardAPIPlugin
+	opts                    *volume.VolumeOptions
+	mounter                 mount.Interface
+}
+
+// This is the spec for the volume that this plugin wraps.
+var wrappedVolumeSpec = &volume.Spec{
+	Name:         "not-used",
+	VolumeSource: api.VolumeSource{EmptyDir: &api.EmptyDirVolumeSource{Medium: api.StorageMediumMemory}},
+}
+
+// SetUp puts in place the volume plugin.
+// This function is not idempotent by design. We want the data to be refreshed periodically.
+// The internal sync interval of kubelet will drive the refresh of data.
+func (m *downwardAPIVolume) SetUp() error {
+	return m.SetUpAt(m.GetPath())
+}
+
+func (m *downwardAPIVolume) SetUpAt(dir string) error {
+	glog.V(3).Infof("Setting up a metadata volume %v for pod %v at %v", m.volName, m.pod.UID, dir)
+	// Wrap EmptyDir,let it do the setup.
+	wrapped, err := m.plugin.host.NewWrapperBuilder(wrappedVolumeSpec, m.pod, *m.opts, m.mounter)
+	if err != nil {
+		return err
+	}
+	if err := wrapped.SetUpAt(dir); err != nil {
+		return err
+	}
+
+	data := make(map[string]string)
+	if err := m.collectData(&data); err != nil {
+		return err
+	}
+
+	// files are fully regenerated only if requested metadata changed
+	if m.isDataChanged(&data) {
+		err = m.writeData(&data)
+	}
+	return err
+}
+
+// IsReadOnly func to fullfill volume.Builder interface
+func (m *downwardAPIVolume) IsReadOnly() bool {
+	return true
+}
+
+// collectData collects requested metadata in data map.
+// Map's key is the requested name of file to dump
+// Map's value is the (sorted) content of the field to be dumped in the file.
+func (m *downwardAPIVolume) collectData(data *map[string]string) error {
+	errlist := []error{}
+	for fieldReference, fileName := range m.fieldReferenceFileNames {
+		if values, err := fieldpath.ExtractFieldPathAsString(m.pod, fieldReference); err != nil {
+			glog.Error(err)
+			errlist = append(errlist, err)
+		} else {
+			(*data)[fileName] = sortLines(values)
+		}
+	}
+	return utilErrors.NewAggregate(errlist)
+}
+
+// isDataChanged iterate over all the entries to check wether at least one
+// file needs to be updated.
+func (m *downwardAPIVolume) isDataChanged(data *map[string]string) bool {
+	for fileName, values := range *data {
+		if isFileToGenerate(path.Join(m.GetPath(), fileName), values) {
+			return true
+		}
+	}
+	return false
+}
+
+// isFileToGenerate compares actual file with the new values. If
+// different (or the file does not exist) return true
+func isFileToGenerate(fileName, values string) bool {
+	if _, err := os.Lstat(fileName); os.IsNotExist(err) {
+		return true
+	}
+	return readFile(fileName) != values
+}
+
+const currentDir = ".current"
+const currentTmpDir = ".current_tmp"
+
+// writeData writes requested metadata in specified files.
+//
+// The file visible in this volume are symlinks to files in the '.current'
```

Comment:
What prevents a user from saying their filename is ".current" or ".current_tmp" or ".current.12345678" ?

You could say leading dots are not allowed, but that limits the utility of this logic if you ever want to use the same technique for secrets or something else.  What if you spec (and validate, and test) that no filename can have a leading ".." and then make all your special files use that?